egg-herbie/main.rkt

```diff
@@ -106,39 +184,53 @@
                       _stdbool ;; simple scheduler?
                       _stdbool ;; constant folding enabled?
                       ->
-                      (iterations : _EGraphIter-pointer)
+                      (iterations : _EGraphIter-pointer) ;; array of _EgraphIter structs
                       ->
-                      (values iterations iterations-length iterations-ptr)))
-
-;; creates a fresh runner from an existing egraph
-(define-eggmath egraph_copy (_fun _egraph-pointer -> _egraph-pointer))
+                      (begin
+                        (define iter-data
+                          (for/list ([i (in-range iterations-length)])
+                            (define ptr (ptr-add iterations i _EGraphIter))
+                            (hash 'nodes
+                                  (EGraphIter-numnodes ptr)
+                                  'eclasses
+                                  (EGraphIter-numeclasses ptr)
+                                  'time
+                                  (EGraphIter-time ptr))))
+                        (destroy_egraphiters iterations-ptr)
+                        iter-data)))
 
 ;; gets the stop reason as an integer
 (define-eggmath egraph_get_stop_reason (_fun _egraph-pointer -> _uint))
 
 ;; egraph -> string
-(define-eggmath egraph_serialize (_fun _egraph-pointer -> _string))
+(define-eggmath egraph_serialize (_fun _egraph-pointer -> _rust/datum))
 
 ;; egraph -> id -> id
 (define-eggmath egraph_find (_fun _egraph-pointer _uint -> _uint))
 
-;; node number -> s-expr string
+;; egraph -> id -> (listof expr)
 (define-eggmath egraph_get_simplest
                 (_fun _egraph-pointer
                       _uint ;; node id
                       _uint ;; iteration
                       ->
-                      _pointer))
+                      _rust/datum)) ;; expr
 
-(define-eggmath egraph_get_proof (_fun _egraph-pointer _string/utf-8 _string/utf-8 -> _pointer))
-
-;; node number -> (s-expr string) string
+;; egraph -> id -> string -> (listof expr)
 (define-eggmath egraph_get_variants
                 (_fun _egraph-pointer
                       _uint ;; node id
-                      _string/utf-8 ;; original expr
+                      _rust/datum ;; original expr
+                      ->
+                      _rust/data)) ;; listof expr
+
+;; egraph -> string -> string -> string
+(define-eggmath egraph_get_proof
+                (_fun _egraph-pointer ;; egraph
+                      _rust/datum ;; expr1
+                      _rust/datum ;; expr2
                       ->
-                      _pointer)) ;; string pointer
+                      _rust/string)) ;; string
```

Comment:
Ok—so there's no regression here, but maybe this comment box is a fine place to relay this thought to you. One issue we have with proofs is that we sometimes make enormously big proofs—megabytes in size. (This is usually from bad merges.) In those cases, just copying and UTF-8 decoding the string can be extremely slow. This is kind of unavoidable if we round-trip through a C string, but in principle it should be possible to have the Rust side tell us how long the string is while it's constructing said string. Then we could skip trying to get the proof if it is too long—we already have the skipping code, it's just inefficiently happening _after_ decoding instead of before.

Reply:
Left comment in code